File: src/cpp/src/memory.cc
#include "horizon_vm/memory.h"
#include "horizon_vm/opcodes.h"
#include <cstring>
#include <stdexcept>

namespace horizon {

namespace {

template <typename T>
T ReadLE(const uint8_t* p) {
  T v;
  std::memcpy(&v, p, sizeof(T));
  return v;
}

template <typename T>
void WriteLE(uint8_t* p, T v) {
  std::memcpy(p, &v, sizeof(T));
}

}  // namespace

Memory::Memory(size_t total_bytes) : physical_(total_bytes, 0) {}

uint32_t Memory::AllocPages(uint32_t count, uint8_t /*flags*/) {
  uint32_t offset = alloc_ptr_;
  uint32_t size   = count * kPageSize;
  if (static_cast<size_t>(offset) + size > physical_.size())
    throw std::runtime_error("Memory: out of physical memory");
  alloc_ptr_ += size;
  return offset;
}

void Memory::MapPage(uint32_t virt_addr, uint32_t phys_offset, uint8_t flags) {
  uint32_t page_num = virt_addr >> 12;
  page_table_[page_num] = {phys_offset, flags};
}

void Memory::UnmapPage(uint32_t virt_addr) {
  page_table_.erase(virt_addr >> 12);
}

uint32_t Memory::Resolve(uint32_t virt_addr, uint8_t access_flags, Ring ring) const {
  uint32_t page_num = virt_addr >> 12;
  uint32_t offset   = virt_addr & 0xFFF;

  auto it = page_table_.find(page_num);
  if (it == page_table_.end() || !(it->second.flags & PAGE_PRESENT)) {
    throw HorizonFault(static_cast<uint8_t>(InterruptVector::PageFault),
                       "Page not mapped: 0x" + std::to_string(virt_addr), virt_addr);
  }
  const auto& entry = it->second;
  if (ring == Ring::User && !(entry.flags & PAGE_USER)) {
    throw HorizonFault(static_cast<uint8_t>(InterruptVector::GPF),
                       "Ring 2 access to kernel page", virt_addr);
  }
  if ((access_flags & PAGE_WRITE) && !(entry.flags & PAGE_WRITE)) {
    throw HorizonFault(static_cast<uint8_t>(InterruptVector::GPF),
                       "Write to read-only page", virt_addr);
  }
  if ((access_flags & PAGE_EXEC) && !(entry.flags & PAGE_EXEC)) {
    throw HorizonFault(static_cast<uint8_t>(InterruptVector::GPF),
                       "Execute on non-exec page", virt_addr);
  }
  return entry.phys_offset + offset;
}

uint8_t Memory::Read8(uint32_t virt, Ring ring) const {
  return physical_[Resolve(virt, PAGE_READ, ring)];
}

uint16_t Memory::Read16(uint32_t virt, Ring ring) const {
  return ReadLE<uint16_t>(&physical_[Resolve(virt, PAGE_READ, ring)]);
}

int32_t Memory::Read32(uint32_t virt, Ring ring) const {
  return ReadLE<int32_t>(&physical_[Resolve(virt, PAGE_READ, ring)]);
}

int64_t Memory::Read64(uint32_t virt, Ring ring) const {
  return ReadLE<int64_t>(&physical_[Resolve(virt, PAGE_READ, ring)]);
}

float Memory::ReadF32(uint32_t virt, Ring ring) const {
  return ReadLE<float>(&physical_[Resolve(virt, PAGE_READ, ring)]);
}

double Memory::ReadF64(uint32_t virt, Ring ring) const {
  return ReadLE<double>(&physical_[Resolve(virt, PAGE_READ, ring)]);
}

void Memory::Write8(uint32_t virt, uint8_t val, Ring ring) {
  physical_[Resolve(virt, PAGE_WRITE, ring)] = val;
}

void Memory::Write16(uint32_t virt, uint16_t val, Ring ring) {
  WriteLE<uint16_t>(&physical_[Resolve(virt, PAGE_WRITE, ring)], val);
}

void Memory::Write32(uint32_t virt, int32_t val, Ring ring) {
  WriteLE<int32_t>(&physical_[Resolve(virt, PAGE_WRITE, ring)], val);
}

void Memory::Write64(uint32_t virt, int64_t val, Ring ring) {
  WriteLE<int64_t>(&physical_[Resolve(virt, PAGE_WRITE, ring)], val);
}

void Memory::WriteF32(uint32_t virt, float val, Ring ring) {
  WriteLE<float>(&physical_[Resolve(virt, PAGE_WRITE, ring)], val);
}

void Memory::WriteF64(uint32_t virt, double val, Ring ring) {
  WriteLE<double>(&physical_[Resolve(virt, PAGE_WRITE, ring)], val);
}

void Memory::WriteBytesPhys(uint32_t phys_offset, const uint8_t* data, size_t len) {
  std::memcpy(&physical_[phys_offset], data, len);
}

const uint8_t* Memory::ReadBytesPhys(uint32_t phys_offset, size_t /*len*/) const {
  return &physical_[phys_offset];
}

}  // namespace horizon

File: src/vm/memory.ts
import { PageFlags, PAGE_SIZE, fault } from './types.js';
import { InterruptVector, Ring } from './opcodes.js';

export interface PageTableEntry {
  physOffset: number; // byte offset into the physical buffer
  flags: number;      // PageFlags bitmask
}

export class Memory {
  private readonly physical: ArrayBuffer;
  private readonly view: DataView;
  private readonly pageTable: Map<number, PageTableEntry> = new Map();
  private allocPtr = 0;

  constructor(totalBytes: number) {
    this.physical = new ArrayBuffer(totalBytes);
    this.view = new DataView(this.physical);
  }

  allocPages(count: number, flags: number): number {
    const physOffset = this.allocPtr;
    const size = count * PAGE_SIZE;
    if (physOffset + size > this.physical.byteLength) {
      throw fault(InterruptVector.GPF, 'Out of physical memory');
    }
    this.allocPtr += size;
    return physOffset;
  }

  mapPage(virtAddr: number, physOffset: number, flags: number): void {
    const pageNum = (virtAddr >>> 0) >>> 12;
    this.pageTable.set(pageNum, { physOffset, flags });
  }

  unmapPage(virtAddr: number): void {
    const pageNum = (virtAddr >>> 0) >>> 12;
    this.pageTable.delete(pageNum);
  }

  private resolve(virtAddr: number, accessFlags: number, ring: Ring): number {
    const vAddr = virtAddr >>> 0;
    const pageNum = vAddr >>> 12;
    const offset = vAddr & 0xFFF;
    const entry = this.pageTable.get(pageNum);

    if (!entry || !(entry.flags & PageFlags.Present)) {
      throw fault(InterruptVector.PageFault, `Page not mapped: 0x${vAddr.toString(16)}`, vAddr);
    }
    if (ring === Ring.User && !(entry.flags & PageFlags.User)) {
      throw fault(InterruptVector.GPF, `Ring 2 access to kernel page: 0x${vAddr.toString(16)}`, vAddr);
    }
    if ((accessFlags & PageFlags.Write) && !(entry.flags & PageFlags.Write)) {
      throw fault(InterruptVector.GPF, `Write to read-only page: 0x${vAddr.toString(16)}`, vAddr);
    }
    if ((accessFlags & PageFlags.Exec) && !(entry.flags & PageFlags.Exec)) {
      throw fault(InterruptVector.GPF, `Execute on non-exec page: 0x${vAddr.toString(16)}`, vAddr);
    }
    return entry.physOffset + offset;
  }

  read8(virt: number, ring: Ring): number {
    const phys = this.resolve(virt, PageFlags.Read, ring);
    return this.view.getUint8(phys);
  }

  read16(virt: number, ring: Ring): number {
    const phys = this.resolve(virt, PageFlags.Read, ring);
    return this.view.getUint16(phys, true);
  }

  read32(virt: number, ring: Ring): number {
    const phys = this.resolve(virt, PageFlags.Read, ring);
    return this.view.getInt32(phys, true);
  }

  read64(virt: number, ring: Ring): bigint {
    const phys = this.resolve(virt, PageFlags.Read, ring);
    return this.view.getBigInt64(phys, true);
  }

  readF32(virt: number, ring: Ring): number {
    const phys = this.resolve(virt, PageFlags.Read, ring);
    return this.view.getFloat32(phys, true);
  }

  readF64(virt: number, ring: Ring): number {
    const phys = this.resolve(virt, PageFlags.Read, ring);
    return this.view.getFloat64(phys, true);
  }

  write8(virt: number, value: number, ring: Ring): void {
    const phys = this.resolve(virt, PageFlags.Write, ring);
    this.view.setUint8(phys, value);
  }

  write16(virt: number, value: number, ring: Ring): void {
    const phys = this.resolve(virt, PageFlags.Write, ring);
    this.view.setUint16(phys, value, true);
  }

  write32(virt: number, value: number, ring: Ring): void {
    const phys = this.resolve(virt, PageFlags.Write, ring);
    this.view.setInt32(phys, value, true);
  }

  write64(virt: number, value: bigint, ring: Ring): void {
    const phys = this.resolve(virt, PageFlags.Write, ring);
    this.view.setBigInt64(phys, value, true);
  }

  writeF32(virt: number, value: number, ring: Ring): void {
    const phys = this.resolve(virt, PageFlags.Write, ring);
    this.view.setFloat32(phys, value, true);
  }

  writeF64(virt: number, value: number, ring: Ring): void {
    const phys = this.resolve(virt, PageFlags.Write, ring);
    this.view.setFloat64(phys, value, true);
  }

  readBytesPhys(physOffset: number, length: number): Uint8Array {
    return new Uint8Array(this.physical, physOffset, length);
  }

  writeBytesPhys(physOffset: number, data: Uint8Array): void {
    new Uint8Array(this.physical, physOffset, data.length).set(data);
  }

  getPhysSize(): number {
    return this.physical.byteLength;
  }
}
